src/Evolution/Systems/CurvedScalarWave/BoundaryConditions/BaylissTurkel.cpp

```diff
@@ -0,0 +1,118 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Systems/CurvedScalarWave/BoundaryConditions/BaylissTurkel.hpp"
+
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <pup.h>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Systems/CurvedScalarWave/Characteristics.hpp"
+#include "Options/ParseOptions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace CurvedScalarWave::BoundaryConditions {
+
+template <size_t Dim>
+std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
+BaylissTurkel<Dim>::get_clone() const noexcept {
+  return std::make_unique<BaylissTurkel>(*this);
+}
+
+template <size_t Dim>
+void BaylissTurkel<Dim>::pup(PUP::er& p) {
+  BoundaryCondition<Dim>::pup(p);
+}
+
+template <size_t Dim>
+BaylissTurkel<Dim>::BaylissTurkel(CkMigrateMessage* const msg) noexcept
+    : BoundaryCondition<Dim>(msg) {}
+
+template <size_t Dim>
+std::optional<std::string> BaylissTurkel<Dim>::dg_time_derivative(
+    const gsl::not_null<Scalar<DataVector>*> dt_pi_correction,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        dt_phi_correction,
+    const gsl::not_null<Scalar<DataVector>*> dt_psi_correction,
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+        face_mesh_velocity,
+    const tnsr::i<DataVector, Dim>& normal_covector,
+    const tnsr::I<DataVector, Dim>& normal_vector,
+    const tnsr::i<DataVector, Dim>& phi, const Scalar<DataVector>& psi,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& coords,
+    const Scalar<DataVector>& gamma1, const Scalar<DataVector>& gamma2,
+    const Scalar<DataVector>& lapse, const tnsr::I<DataVector, Dim>& shift,
+    const Scalar<DataVector>& dt_pi, const tnsr::i<DataVector, Dim>& dt_phi,
+    const Scalar<DataVector>& dt_psi, const tnsr::i<DataVector, Dim>& d_psi,
+    const tnsr::ij<DataVector, Dim>& d_phi) const noexcept {
+  const DataVector inv_radius = 1. / get(magnitude(coords));
```

Comment:
maybe use a `Variables` to reduce the number of allocations in this function? There are currently 6 I'm seeing immediately

Reply:
hm, I could only spot 3. I have removed 2 of them.

Reply:
Ah, I should've listed what I saw, sorry! I see:
- 4 from `tnsr::a<DataVector, 3, Frame::Inertial> char_speeds{lapse.size()};`
- 1 from `face_speed = get(dot_product(normal_covector, *face_mesh_velocity));`
- I think you got the other one :)

Please feel free to squash this :)

Reply:
ah, I had not realized that Tensors actually just forward the allocation to the underlying `std::array` - have we ever considered changing the tensor constructor to always make a single allocation?

Reply:
We've talked about this and this was originally the plan. It hasn't been prioritized partly because in most places we need several `Tensor` objects and so `Variables` are a better solution and partly because making sure all the semantics are correct will be tricky. It took us 3 months of using `DataVector`s before we settled on an implementation that is reasonably intuitive. So it certainly could be done, we just haven't had enough cases where it outweighs using a `Variables` to justify the development effort.